libraries/previousAuctionInfo/previousAuctionInfo.js

```diff
@@ -0,0 +1,130 @@
+import {on as onEvent} from '../../src/events.js';
+import { EVENTS } from '../../src/constants.js';
+
+export let previousAuctionInfoEnabled = false;
+let enabledBidders = [];
+export let winningBidsMap = {};
+
+export const resetPreviousAuctionInfo = () => {
+  previousAuctionInfoEnabled = false;
+  enabledBidders = [];
+  winningBidsMap = {};
+};
+
+export const enablePreviousAuctionInfo = (config, cb = initHandlers) => {
+  const { bidderCode, isBidRequestValid } = config;
+  const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderCode);
+  if (!enabledBidder) enabledBidders.push({ bidderCode, isBidRequestValid, maxQueueLength: config.maxQueueLength || 10 });
+  if (previousAuctionInfoEnabled) return;
+  previousAuctionInfoEnabled = true;
+  cb();
+}
+
+export const initHandlers = () => {
+  onEvent(EVENTS.AUCTION_END, onAuctionEndHandler);
+  onEvent(EVENTS.BID_WON, onBidWonHandler);
+  onEvent(EVENTS.BID_REQUESTED, onBidRequestedHandler);
+};
+
+export const onAuctionEndHandler = (auctionDetails) => {
+  // eslint-disable-next-line no-console
+  console.log('onAuctionEndHandler', auctionDetails);
+
+  try {
+    let highestCpmBid = 0;
+    const receivedBidsMap = {};
+    const rejectedBidsMap = {};
+
+    if (auctionDetails.bidsReceived && auctionDetails.bidsReceived.length) {
+      highestCpmBid = auctionDetails.bidsReceived.reduce((highestBid, currentBid) => {
+        return currentBid.cpm > highestBid.cpm ? currentBid : highestBid;
+      }, auctionDetails.bidsReceived[0]);
+
+      auctionDetails.bidsReceived.forEach(bidReceived => {
+        receivedBidsMap[bidReceived.requestId] = bidReceived;
+      });
+    }
+
+    if (auctionDetails.bidsRejected && auctionDetails.bidsRejected.length) {
+      auctionDetails.bidsRejected.forEach(bidRejected => {
+        rejectedBidsMap[bidRejected.requestId] = bidRejected;
+      });
+    }
+
+    if (auctionDetails.bidderRequests && auctionDetails.bidderRequests.length) {
+      auctionDetails.bidderRequests.forEach(bidderRequest => {
+        const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderRequest.bidderCode);
+
+        if (enabledBidder) {
+          bidderRequest.bids.forEach(bid => {
+            const isValidBid = enabledBidder.isBidRequestValid(bid);
+
+            if (!isValidBid) return;
+
+            const previousAuctionInfoPayload = {
+              bidderRequestId: bidderRequest.bidderRequestId,
+              minBidToWin: highestCpmBid?.cpm || 0,
+              rendered: 0,
+              transactionId: bid.ortb2Imp.ext.tid || bid.transactionId,
+              source: 'pbjs',
+              auctionId: auctionDetails.auctionId,
+              impId: bid.adUnitCode,
```

Comment:
IMO `adUnitCode` is a better name. `impId` makes me think of ORTB `imp.id`, which most of the time is different from the ad unit code.